### PR TITLE
feat(Effects): Add root effects init action

### LINKED ISSUE #246: The way to fire an effect after effects are settled up
## I'm submitting a...

<pre><code>
[x] Regression (a behavior that used to work and stopped working in a new release)
[ ] Bug report  
[ ] Feature request
[ ] Documentation issue or request
[ ] Support request
</code></pre>

## What is the current behavior?

There is no action, no function which fired when all effects are settled up. Where we can dispatch actions to fire other effects.

## Expected behavior:

```typescript
    this.onInit$ = this.actions$
      .ofType('@ngrx/effects/up')
      .switchMap((isAuthorized) => {
        return this.isAuthorized$;
      })
      .map((isAuthorized) => {
        if (isAuthorized) {
          this.store.dispatch({ type: '[posts] getAll' })
        }
      });

    @Effect()
    this.onGetPosts$ = this.actions$
      .ofType('[posts] getAll')
      .map(() => ... )
```
or, may be
```typescript
  public ngrxOnRunEffects(resolvedEffects$: Observable<EffectNotification>): Observable<EffectNotification> {
    this.isAuthorized$
      .take(1)
      .subscribe({
        next: (isAuthorized) => {
          if (isAuthorized) {
            this.store.dispatch({ type: '[posts] getAll' })
          }
        },
      });
    return resolvedEffects$;
  }
```

## Minimal reproduction of the problem with instructions:

There a few issues, of `@ngrx/store/init` action. Seems, that this action is not useful for described purpose anymore. So, we probably need another.
#103 
#152 

## Version of affected browser(s),operating system(s), npm, node and ngrx:
ngrx v4

## Other information:

Possible workaround:

dispatch action, somewhere after application start.
In the `AppComponent` for example.
```typescript
this.store.dispatch({ type: '[app] init' });
```
And use it to dispatch other actions, that must be dispatched after all effects are settled up.

### PR DESCRIPTION
**Pull request by tdeschryver**
Closes #246

### CODE REVIEW

File: modules/effects/src/effects_root_module.ts [hunk 1]
```diff
@@ -1,18 +1,22 @@
...
 export class EffectsRootModule {
   constructor(
     private sources: EffectSources,
     runner: EffectsRunner,
+    store: Store<any>,
     @Inject(ROOT_EFFECTS) rootEffects: any[],
     @Optional() storeModule: StoreModule
   ) {
     runner.start();
+    store.dispatch({ type: ROOT_EFFECTS_INIT });
```
Comment (MikeRyanDev) [inline modules/effects/src/effects_root_module.ts]: I think this needs to be beneath the lines where you add the effects.

File: modules/effects/src/effects_root_module.ts [hunk 2]
```diff
@@ -1,18 +1,22 @@
 import { NgModule, Inject, Optional } from '@angular/core';
-import { StoreModule } from '@ngrx/store';
+import { StoreModule, Store } from '@ngrx/store';
 import { EffectsRunner } from './effects_runner';
 import { EffectSources } from './effect_sources';
 import { ROOT_EFFECTS } from './tokens';
 
+export const ROOT_EFFECTS_INIT = '@ngrx/store/root effects init' as '@ngrx/store/root effects init';
```
Comment (brandonroberts) [inline modules/effects/src/effects_root_module.ts]: I would rename the type string to `@ngrx/effects/init`

File: modules/effects/src/effects_root_module.ts (hunk 1 above)
Comment (tdeschryver) [reply modules/effects/src/effects_root_module.ts]: Changed.
Why exactly :smile: ?

File: modules/effects/src/effects_root_module.ts (hunk 2 above)
Comment (tdeschryver) [reply modules/effects/src/effects_root_module.ts]: Yep you're right, don't know what I was thinking!

File: modules/effects/src/effects_root_module.ts (hunk 1 above)
Comment (MikeRyanDev) [reply modules/effects/src/effects_root_module.ts]: Dispatching and registering the effects happens synchronously. Moving the dispatch to after registering the effects will give the effects a chance to listen for this new action type.

**Comment by MikeRyanDev** [comment]
Thanks @tdeschryver!

**Comment by felikf** [comment]
I am just wondering ... is this supposed to work in feature effects (`EffectsModule.forFeature([Effects]`)? Because it does not...

If I use:

    @Effect({ dispatch: false }) init$= this.actions$
        .ofType(ROOT_EFFECTS_INIT)
        .do(console.log.bind(console, 'HIT!!!!'));

in root module, it works.

When I use it in feature module, it does not.

I would need this for feature module though... :)

**Pull request merged**